src/network/networkd-routing-policy-rule.c

```diff
@@ -1235,18 +1235,40 @@ int routing_policy_load_rules(const char *state_file, Set **rules) {
 }
 
 void routing_policy_rule_purge(Manager *m, Link *link) {
-        RoutingPolicyRule *rule, *existing;
-        Iterator i;
+        RoutingPolicyRule *rule, *existing, *link_rule;
+        Iterator i, j;
+        Link *l;
         int r;
 
         assert(m);
         assert(link);
 
         SET_FOREACH(rule, m->rules_saved, i) {
+                bool found = false;
+
                 existing = set_get(m->rules_foreign, rule);
                 if (!existing)
                         continue; /* Saved rule does not exist anymore. */
 
+                HASHMAP_FOREACH(l, m->links, j) {
+                        if (!link->network)
```

Comment:
This is wrong — `l` is never used.

I think it'd make a lot of sense to split out this internal loop over links into a separate function, `manager_links_have_routing_rule(Manager *m, const Rule *rule)` or smth. Then `found` will not have to be defined because you can return `true` or `false` directly.

Reply:
Ouch!